### PR TITLE
convert testFailedDeployment e2e test to integration test

### PR DESCRIPTION
**Pull request by crimsonfaith91**
**What this PR does / why we need it**:
This PR convert a deployment e2e test named "testFailedDeployment" to integration test.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: xref #52113 

**Release note**:

```release-note
NONE
```
/assign

### CODE REVIEW

File: test/utils/deployment.go [hunk 1]
```diff
@@ -267,3 +267,43 @@ func WaitForDeploymentCompletes(c clientset.Interface, d *extensions.Deployment,
...
+		if err != nil {
+			return false, err
+		}
+		if deployment.Status.ObservedGeneration >= desiredGeneration && deployment.Status.UpdatedReplicas >= minUpdatedReplicas {
+			return true, nil
+		}
+		return false, nil
+	})
+	if err != nil {
+		return fmt.Errorf("error waiting for deployment %s to have at least %d updpatedReplicas: %v", deploymentName, minUpdatedReplicas, err)
```
Comment (crimsonfaith91) [inline test/utils/deployment.go]: `updatedReplicas`

File: test/utils/deployment.go [hunk 2]
```diff
@@ -267,3 +267,43 @@ func WaitForDeploymentCompletes(c clientset.Interface, d *extensions.Deployment,
...
+	}
+	return nil
+}
+
```
Comment (crimsonfaith91) [inline test/utils/deployment.go]: Add a comment about what this function does.

File: test/integration/deployment/deployment_test.go [hunk 3]
```diff
@@ -498,3 +498,48 @@ func TestRollbackDeploymentRSNoRevision(t *testing.T) {
...
+	if err = tester.waitForDeploymentWithCondition(deploymentutil.TimedOutReason, v1beta1.DeploymentProgressing); err != nil {
+		t.Fatalf("expecting deployment %q has %q condition type with %q reason in .status, but found none: %v", deploymentName, v1beta1.DeploymentProgressing, deploymentutil.TimedOutReason, err)
+	}
+
+	// Verify deployment status becomes valid while manually marking pods as ready simultaneously
+	if err := tester.waitForDeploymentCompleteAndMarkPodsReady(); err != nil {
+		t.Fatalf("deployment %q fails to have its status becoming valid: %v", deploymentName, err)
+	}
+
+	// Verify deployment status has Progressing condition type with NewReplicaSetAvailable reason after the pods are marked ready
```
Comment (crimsonfaith91) [inline test/integration/deployment/deployment_test.go]: @janetkuo Do i need to add other step(s)? The test looks good to me so far.

File: test/utils/deployment.go [hunk 4]
```diff
@@ -301,3 +301,43 @@ func WaitForDeploymentRollbackCleared(c clientset.Interface, ns, deploymentName
...
+	})
+	if err != nil {
+		return fmt.Errorf("error waiting for deployment %s to have at least %d updatedReplicas: %v", deploymentName, minUpdatedReplicas, err)
+	}
+	return nil
+}
+
+func WaitForDeploymentWithCondition(c clientset.Interface, ns, deploymentName, reason string, condType extensions.DeploymentConditionType, logf LogfFn, pollInterval, pollTimeout time.Duration) error {
+	var deployment *extensions.Deployment
+	pollErr := wait.PollImmediate(time.Second, 5*time.Minute, func() (bool, error) {
```
Comment (janetkuo) [inline test/utils/deployment.go]: use the input poll interval and timeout

File: test/utils/deployment.go [hunk 5]
```diff
@@ -301,3 +301,43 @@ func WaitForDeploymentRollbackCleared(c clientset.Interface, ns, deploymentName
...
+		d, err := c.Extensions().Deployments(ns).Get(deploymentName, metav1.GetOptions{})
```
Comment (janetkuo) [inline test/utils/deployment.go]: use explicit version

File: test/integration/deployment/deployment_test.go [hunk 6]
```diff
@@ -498,3 +498,48 @@ func TestRollbackDeploymentRSNoRevision(t *testing.T) {
...
+
+	// Start informer and controllers
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	informers.Start(stopCh)
+	go rm.Run(5, stopCh)
+	go dc.Run(5, stopCh)
+
+	if err = tester.waitForDeploymentUpdatedReplicasLTE(replicas); err != nil {
+		t.Fatalf("deployment %q failed to come up with %d replicas (with non-ready pods)", deploymentName, replicas)
```
Comment (janetkuo) [inline test/integration/deployment/deployment_test.go]: `err` should be printed too. However, `waitForDeploymentUpdatedReplicasLTE()` already returns similar message here, so returning raw error would be enough.

File: test/utils/deployment.go [hunk 7]
```diff
@@ -301,3 +301,43 @@ func WaitForDeploymentRollbackCleared(c clientset.Interface, ns, deploymentName
...
+		if err != nil {
+			return false, err
+		}
+		if deployment.Status.ObservedGeneration >= desiredGeneration && deployment.Status.UpdatedReplicas >= minUpdatedReplicas {
+			return true, nil
+		}
+		return false, nil
+	})
+	if err != nil {
+		return fmt.Errorf("error waiting for deployment %s to have at least %d updatedReplicas: %v", deploymentName, minUpdatedReplicas, err)
```
Comment (janetkuo) [inline test/utils/deployment.go]: suggest including `deployment.Status.UpdatedReplicas` in the returned error message too

File: test/integration/deployment/deployment_test.go [hunk 8]
```diff
@@ -498,3 +498,48 @@ func TestRollbackDeploymentRSNoRevision(t *testing.T) {
...
+	}
+
+	// Verify deployment status has Progressing condition type with ProgressDeadlineExceeded reason
```
Comment (janetkuo) [inline test/integration/deployment/deployment_test.go]: Document more details here:

```go
// Pods are not marked as Ready, therefore the deployment progress will eventually timeout after progressDeadlineSeconds has passed.
// Wait for the deployment to have a progress timeout condition.
```

File: test/integration/deployment/deployment_test.go [hunk 9]
```diff
@@ -498,3 +498,48 @@ func TestRollbackDeploymentRSNoRevision(t *testing.T) {
...
+	if err = tester.waitForDeploymentWithCondition(deploymentutil.TimedOutReason, v1beta1.DeploymentProgressing); err != nil {
+		t.Fatalf("expecting deployment %q has %q condition type with %q reason in .status, but found none: %v", deploymentName, v1beta1.DeploymentProgressing, deploymentutil.TimedOutReason, err)
+	}
+
+	// Verify deployment status becomes valid while manually marking pods as ready simultaneously
```
Comment (janetkuo) [inline test/integration/deployment/deployment_test.go]:
```go
// Manually mark pods as Ready and wait for deployment to complete.
```

File: test/integration/deployment/deployment_test.go (hunk 3 above)
Comment (janetkuo) [inline test/integration/deployment/deployment_test.go]: // Wait for the deployment to have a progress complete condition.

File: test/utils/deployment.go [hunk 10]
```diff
@@ -301,3 +301,43 @@ func WaitForDeploymentRollbackCleared(c clientset.Interface, ns, deploymentName
...
+		d, err := c.Extensions().Deployments(ns).Get(deploymentName, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		deployment = d
+		cond := deploymentutil.GetDeploymentCondition(deployment.Status, condType)
+		return cond != nil && cond.Reason == reason, nil
+	})
+	if pollErr == wait.ErrWaitTimeout {
+		pollErr = fmt.Errorf("deployment %q never updated with the desired condition and reason: %v", deployment.Name, deployment.Status.Conditions)
```
Comment (janetkuo) [inline test/utils/deployment.go]: ": %v" to ", latest deployment conditions: %+v"

File: test/integration/deployment/deployment_test.go [hunk 11]
```diff
@@ -498,3 +498,48 @@ func TestRollbackDeploymentRSNoRevision(t *testing.T) {
...
+	go rm.Run(5, stopCh)
+	go dc.Run(5, stopCh)
+
+	if err = tester.waitForDeploymentUpdatedReplicasLTE(replicas); err != nil {
+		t.Fatalf("deployment %q failed to come up with %d replicas (with non-ready pods)", deploymentName, replicas)
+	}
+
+	// Verify deployment status has Progressing condition type with ProgressDeadlineExceeded reason
+	if err = tester.waitForDeploymentWithCondition(deploymentutil.TimedOutReason, v1beta1.DeploymentProgressing); err != nil {
+		t.Fatalf("expecting deployment %q has %q condition type with %q reason in .status, but found none: %v", deploymentName, v1beta1.DeploymentProgressing, deploymentutil.TimedOutReason, err)
```
Comment (janetkuo) [inline test/integration/deployment/deployment_test.go]: err returned from `waitForDeploymentWithCondition()` has enough information, just return it is enough

File: test/integration/deployment/deployment_test.go [hunk 12]
```diff
@@ -498,3 +498,48 @@ func TestRollbackDeploymentRSNoRevision(t *testing.T) {
...
+	}
+
+	// Verify deployment status becomes valid while manually marking pods as ready simultaneously
+	if err := tester.waitForDeploymentCompleteAndMarkPodsReady(); err != nil {
+		t.Fatalf("deployment %q fails to have its status becoming valid: %v", deploymentName, err)
+	}
+
+	// Verify deployment status has Progressing condition type with NewReplicaSetAvailable reason after the pods are marked ready
+	if err = tester.waitForDeploymentWithCondition(deploymentutil.NewRSAvailableReason, v1beta1.DeploymentProgressing); err != nil {
+		t.Fatalf("expecting deployment %q has %q condition type with %q reason in .status, but found none: %v", deploymentName, v1beta1.DeploymentProgressing, deploymentutil.NewRSAvailableReason, err)
```
Comment (janetkuo) [inline test/integration/deployment/deployment_test.go]: ditto, returning err is enough

File: test/utils/deployment.go (hunk 10 above)
Comment (janetkuo) [reply test/utils/deployment.go]: In the returned error message, can also include the information about the condition & reason the caller is waiting for

File: test/integration/deployment/deployment_test.go [hunk 13]
```diff
@@ -498,3 +498,48 @@ func TestRollbackDeploymentRSNoRevision(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+func TestFailedDeployment(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (crimsonfaith91) [reply test/integration/deployment/deployment_test.go]: sure!

File: test/utils/deployment.go (hunk 4 above)
Comment (crimsonfaith91) [reply test/utils/deployment.go]: oops, totally missed them!

File: test/utils/deployment.go (hunk 5 above)
Comment (crimsonfaith91) [reply test/utils/deployment.go]: oh, i just copied the function without checking the details within.

**Comment by janetkuo** [comment]
Here we list all pods selected by the deployment and expect the number of pods to match deployment replicas. However, since we have two overlapping deployments in this test, you'd need to update here to exclude the pods that aren't owned by the input deployment:
https://github.com/kubernetes/kubernetes/blob/aabfc5555b611589996efc25cba4a679d2d2f044/test/integration/deployment/util.go#L219-L227

**Review by janetkuo — APPROVED** [review]

**Comment by janetkuo** [comment]
lgtm, please squash commits

**Pull request merged**